### PR TITLE
`Programming exercises`: Highlight code issues in short result information

### PR DESCRIPTION
**Pull request by b-fein**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context
PR #6460 changed that the result string appear by default in the short format. It might be the case, though, that a student has passed most/all test cases, but still has code style/static code analysis issues that should be resolved. The short format gives does not indicate that in any way without clicking on it to inspect the detailed feedback.

### Description
To indicate that code issues exist while keeping the result string short, a warning icon is added to the short format of the result string.
In the long format no icon is used since the string already indicates the concrete number of issues.

The icon shows a tooltip on hover that mentions the code analysis warnings. With that combined with the identical colour the necessary mental connection between this part of the result string and the relevant section (‘Warnings’) in the detailed feedback should be clear for students.
 
### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Programming Exercise with static code analysis

1. Log in to Artemis
2. Edit the solution participation code to introduce a static code analysis issue.
3. All tests pass, the result shows 100%, but the result string contains the warning icon as shown in the screenshots below. On hover over the icon, a translated tooltip is shown.

#### Performance Review
- [x] I confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| `result.component.ts` | 100% for the relevant cases in the HTML template | ✅              |

### Screenshots

Warning icon in case of SCA issues regardless of score:
grafik
grafik
grafik
Long result variant has no icon since the text already provides the information:
Screenshot 2023-05-17 at 09-05-18 Programming Exercises

<details>
<summary>Iteration 1: Colour change</summary>

With issues:
image
Screenshot 2023-05-15 at 17-04-58 Code Editor

Without issues:
image

0% still red:
image
</details>

### CODE REVIEW

**Review by DominikRemo — APPROVED** [review]
Tested on ts2, works as expected

**Review by SolizerCodes — DISMISSED** [review]
Tested on ts1. Everything works as expected.

**Review by dakennguyen — APPROVED** [review]
tested on ts1

**Review by eceeeren — DISMISSED** [review]
Tested on ts2, works as expected

**Review by greekgeek1 — DISMISSED** [review]
Tested on ts3. Works as expected.

**Comment by krusche** [comment]
I am not sure this is a good idea.
I think we have the convention to change the color based on the percentage. If I remember correctly, we have the following mapping
< 40%: red
40-80%: yellow
> 80%: green

Now if you suddenly show 100% as yellow, this might be confusing.

I understand the issue that 100% green can hide code issues, however I would like to use a consistent color scheme to adapt the scheme to


One potential solution might be:

< 50%: red
50-100%: yellow (at least one failing test or code issue even if the weight is 0)
100%: green (everything passes)

Notice, the score might also be larger than 100% in case there are bonus points.

We might want to implement the logic for determining the color on the server to avoid multiple implementations.

Let's discuss such changes quickly with the maintainers and developers, before directly implementing them

**Comment by b-fein** [comment]
> One potential solution might be:
>
> < 50%: red
> 50-100%: yellow (at least one failing test or code issue even if the weight is 0)
> 100%: green (everything passes)

That is pretty much exactly how it is implemented. There is no special case for ‘100% with code issues’. The score interval always takes precedence. Instead:

- [0, 40%): _always_ red
- [40%, 80%): _always_ orange
- [80%, ∞):
  - if code issues exist: orange
  - if no code issues exist: green
 
That way orange consistently indicates that there is room for improvement in the submission.

**Comment by krusche** [comment]
> > One potential solution might be:
> > < 50%: red
> > 50-100%: yellow (at least one failing test or code issue even if the weight is 0)
> > 100%: green (everything passes)
> 
> That is pretty much exactly how it is implemented. There is no special case for ‘100% with code issues’. The screenshots might not have made that clear enough. The score interval always takes precedence. Instead:
> 
> * [0, 40%): _always_ red
> * [40%, 80%): _always_ orange
> * [80%, ∞):
>   
>   * if code issues exist: orange
>   * if no code issues exist: green
> 
> That way orange consistently indicates that there is room for improvement in the submission, either by fixing test cases to increase the score, or by fixing SCA issues.

I still find this confusing, because a student with 81% and no code issues would have a green color, while a student with 100% and one small code issue (which is not even influencing the score) would have an orange color. I don't think this would be a good solution.

**Comment by krusche** [comment]
would it be a solution to show a yellow warning icon next to the score in case code issues exist? This could be done independently of the actual score. I think this would be way more intuitive

Examples:

* 100% ⚠️ (a few seconds ago) --> text in green, with code issues
* 100% (a few seconds ago) --> text in green, without code issues
* 75% ⚠️ (a few seconds ago) --> text in orange, with code issues
* 75% (a few seconds ago) --> text in orange, without code issues

**Comment by b-fein** [comment]
> Would it be a solution to show a yellow warning icon next to the score in case code issues exist?

That is a good solution to keep the result string short without losing information. I like that idea. :+1:

**Comment by b-fein** [comment]
Implemented this second iteration and updated the screenshots.

File: src/main/webapp/app/exercises/shared/result/result.component.html [hunk 1]
```diff
@@ -19,12 +19,13 @@
                 id="result-score"
                 (click)="showDetails(result)"
             >
-                <span class="guided-tour result" [class.clickable-result]="exercise?.type !== ExerciseType.QUIZ" [ngbTooltip]="resultTooltip | artemisTranslate">
+                <span class="guided-tour result" [ngbTooltip]="resultTooltip | artemisTranslate">
                     {{ resultString }}
                 </span>
-                <span class="result" [class.clickable-result]="exercise?.type !== ExerciseType.QUIZ" [ngbTooltip]="result.completionDate | artemisDate">
-                    ({{ result.completionDate | artemisTimeAgo }})
+                <span class="result" *ngIf="short && (result?.codeIssueCount ?? 0) > 0" ngbTooltip="{{ 'artemisApp.result.codeIssuesTooltip' | artemisTranslate }}">
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/shared/result/result.component.html:25]:
```suggestion
                <span class="result" *ngIf="short && result?.codeIssueCount" ngbTooltip="{{ 'artemisApp.result.codeIssuesTooltip' | artemisTranslate }}">
```

File: src/main/webapp/i18n/de/result.json [hunk 2]
```diff
@@ -84,6 +84,7 @@
             "preliminary": "vorläufig",
             "preliminaryTooltip": "Dein Ergebnis ist noch nicht endgültig, weil weitere Tests nach der Einreichungsfrist ausgeführt werden.",
             "preliminaryTooltipSemiAutomatic": "Dein Ergebnis ist noch nicht endgültig, weil weitere Tests nach der Einreichungsfrist ausgeführt werden oder eine manuelle Bewertung aussteht.",
+            "codeIssuesTooltip": "Die automatische Codeanalyse hat mögliche Probleme in deinem Code gefunden.",
```
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/result.json:87]:
```suggestion
            "codeIssuesTooltip": "Die automatische Codeanalyse hat mögliche Probleme in deinem Codestil gefunden.",
```
Or
```suggestion
            "codeIssuesTooltip": "Die automatische Codeanalyse hat Codeissues gefunden.",
```
"Probleme" might sound a bit to alarming to some students

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
Looks good overall, left two comments:

File: src/main/webapp/i18n/de/result.json (hunk 2 above)
Comment (b-fein) [reply src/main/webapp/i18n/de/result.json]: Sure, ‘Codeissues’ sounds good.

**Review by SolizerCodes — APPROVED** [review]
Tested on ts1, the "warning icon" is displayed correctly. Colors are correct, too.

**Review by JohannesStoehr — APPROVED** [review]
Thanks for addressing my feedback. Code looks good to me

**Review by DominikRemo — APPROVED** [review]
Tested on ts2. Works great!

**Review by undernagruzez — COMMENTED** [review]
Tested on ts5, works as expected.

**Review by chrisknedl — APPROVED** [review]
Had a look at the code changes, and they look good.

**Pull request merged**